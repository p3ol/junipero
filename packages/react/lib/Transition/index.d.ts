import { ReactNode, ComponentPropsWithRef } from 'react';

export declare interface TransitionProps extends ComponentPropsWithRef<any> {
  children?: ReactNode | JSX.Element;
  in: boolean;
  mounterOnEnter?: boolean;
  name?: string;
  timeout?: number | { enter: number; exit: number };
  unmountOnExit?: boolean;
  onEnter?(): void;
  onEntering?(): void;
  onEntered?(): void;
  onExit?(): void;
  onExiting?(): void;
  onExited?(): void;
}

declare function Transition(props: TransitionProps): ReactNode | JSX.Element;

export default Transition;
