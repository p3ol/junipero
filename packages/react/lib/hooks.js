import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
