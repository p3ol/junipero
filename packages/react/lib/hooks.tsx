import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
