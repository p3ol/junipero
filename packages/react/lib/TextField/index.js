import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
