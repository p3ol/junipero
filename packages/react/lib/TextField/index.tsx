import {
  type ChangeEvent,
  type FocusEvent,
  type RefObject,
  type WheelEvent,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, exists } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import { mockState } from '../utils';
import { useFieldControl } from '../hooks';

export declare interface TextFieldRef extends JuniperoRef {
  dirty: boolean;
  focused: boolean;
  valid: boolean;
  value: string | number;
  blur(): void;
  focus(): void;
  reset(): void;
  setDirty(dirty: boolean): void;
  innerRef: RefObject<HTMLDivElement>;
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
}

export declare interface TextFieldProps
  extends SpecialComponentPropsWithRef<any, TextFieldRef> {
  autoFocus?: boolean;
  disabled?: boolean;
  required?: boolean;
  type?: string;
  valid?: boolean;
  value?: string | number;
  onBlur?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onChange?(field: FieldContent<string | number>): void;
  onFocus?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onWheel?(event: WheelEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onValidate?(
    val: string | number,
    flags: { required: boolean; dirty: boolean }
  ): boolean;
}

export declare interface TextFieldState {
  value: string | number;
  valid: boolean;
  dirty: boolean;
  focused: boolean;
}

const TextField = ({
  ref,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val: string | number, opts?: { required: boolean }) =>
    !!val || !opts?.required,
  ...rest
}: TextFieldProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>(null);
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState<TextFieldState>, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = (value?: string) => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = (dirty: boolean) => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? ((_: WheelEvent) => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
};

TextField.displayName = 'TextField';

export default TextField;
