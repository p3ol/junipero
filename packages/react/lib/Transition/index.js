import { Children, useState, useCallback, cloneElement } from 'react';
import { useTimeout, useLayoutEffectAfterMount } from '@junipero/hooks';
import { classNames } from '@junipero/core';
import PropTypes from 'prop-types';

const UNMOUNTED = 'unmounted';
const ENTER = 'enter';
const EXIT = 'exit';
const IDLE = 'idle';
const STARTING = 'starting';
const ACTIVE = 'active';
const DONE = 'done';

const Transition = ({
  children,
  in: inProp,
  name = 'transition',
  timeout = 100,
  mountOnEnter = true,
  unmountOnExit = false,
  ...rest
}) => {
  const [status, setStatus] = useState(inProp ? ENTER : UNMOUNTED);
  const [step, setStep] = useState(inProp ? STARTING : IDLE);

  useLayoutEffectAfterMount(() => {
    setStatus(inProp ? ENTER : EXIT);
    setStep(STARTING);
  }, [inProp]);

  useTimeout(() => {
    setStep(ACTIVE);
  }, 0, [step], { enabled: step === STARTING });

  useTimeout(() => {
    if (step !== IDLE) {
      setStep(DONE);
    }
  }, timeout?.enter ?? timeout, [status, step], {
    enabled: status === ENTER,
  });

  useTimeout(() => {
    if (step !== IDLE) {
      setStatus(UNMOUNTED);
      setStep(DONE);
    }
  }, timeout?.exit ?? timeout, [status, step], {
    enabled: status === EXIT,
  });

  const getClassName = useCallback(() => (
    name + '-' + status + (![IDLE, STARTING].includes(step) ? '-' + step : '')
  ), [status, step]);

  const child = Children.only(children);

  return status !== UNMOUNTED && (!unmountOnExit || mountOnEnter)
    ? cloneElement(child, {
      className: classNames(child.props.className, getClassName()),
      ...rest,
    }) : null;
};

Transition.displayName = 'Transition';
Transition.propTypes = {
  in: PropTypes.bool,
  name: PropTypes.string,
  timeout: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      enter: PropTypes.number,
      exit: PropTypes.number,
    }),
  ]),
  mountOnEnter: PropTypes.bool,
  unmountOnExit: PropTypes.bool,
};

export default Transition;
