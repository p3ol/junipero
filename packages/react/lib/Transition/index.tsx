import {
  type ComponentPropsWithoutRef,
  type ReactElement,
  cloneElement,
  useState,
  useRef,
  useCallback,
} from 'react';
import { useTimeout, useLayoutEffectAfterMount } from '@junipero/hooks';
import { classNames } from '@junipero/core';

export const TRANSITION_STATE_UNMOUNTED = 'unmounted';
export const TRANSITION_STATE_ENTER = 'enter';
export const TRANSITION_STATE_EXIT = 'exit';
export const TRANSITION_STATE_IDLE = 'idle';
export const TRANSITION_STATE_STARTING = 'starting';
export const TRANSITION_STATE_ACTIVE = 'active';
export const TRANSITION_STATE_DONE = 'done';

export declare interface TransitionTimeoutObject {
  enter?: number;
  exit?: number;
}

export declare interface TransitionProps
  extends ComponentPropsWithoutRef<any> {
  in: boolean;
  mounterOnEnter?: boolean;
  name?: string;
  timeout?: number | TransitionTimeoutObject;
  unmountOnExit?: boolean;
  onEnter?(): void;
  onEntering?(): void;
  onEntered?(): void;
  onExit?(): void;
  onExiting?(): void;
  onExited?(): void;
}

const Transition = ({
  children,
  in: inProp,
  name = 'transition',
  timeout = 100,
  mountOnEnter = true,
  unmountOnExit = false,
  onEnter,
  onEntering,
  onEntered,
  onExit,
  onExiting,
  onExited,
  ...rest
}: TransitionProps) => {
  const previousIn = useRef(inProp);
  const [status, setStatus] = useState(
    inProp ? TRANSITION_STATE_ENTER : TRANSITION_STATE_UNMOUNTED
  );
  const [step, setStep] = useState(
    inProp ? TRANSITION_STATE_STARTING : TRANSITION_STATE_IDLE
  );

  useLayoutEffectAfterMount(() => {
    if (inProp === previousIn.current) {
      return;
    }

    previousIn.current = inProp;
    setStatus(inProp ? TRANSITION_STATE_ENTER : TRANSITION_STATE_EXIT);

    if (inProp) {
      onEnter?.();
    } else {
      onExit?.();
    }

    setStep(TRANSITION_STATE_STARTING);
  }, [inProp]);

  useTimeout(() => {
    setStep(TRANSITION_STATE_ACTIVE);

    if (status === TRANSITION_STATE_ENTER) {
      onEntering?.();
    } else {
      onExiting?.();
    }
  }, 0, [step, status], { enabled: step === TRANSITION_STATE_STARTING });

  useTimeout(() => {
    if (step !== TRANSITION_STATE_IDLE) {
      setStep(TRANSITION_STATE_DONE);
      onEntered?.();
    }
  },
  (timeout as TransitionTimeoutObject)?.enter ?? (timeout as number),
  [status, step],
  { enabled: status === TRANSITION_STATE_ENTER });

  useTimeout(() => {
    if (step !== TRANSITION_STATE_IDLE) {
      if (unmountOnExit) {
        setStatus(TRANSITION_STATE_UNMOUNTED);
      }

      setStep(TRANSITION_STATE_DONE);
      onExited?.();
    }
  },
  (timeout as TransitionTimeoutObject)?.exit ?? (timeout as number),
  [status, step],
  { enabled: status === TRANSITION_STATE_EXIT });

  const getClassName = useCallback(() => (
    status === TRANSITION_STATE_UNMOUNTED ? '' : (
      name + '-' + status + (
        ![TRANSITION_STATE_IDLE, TRANSITION_STATE_STARTING].includes(step)
          ? '-' + step
          : ''
      )
    )
  ), [status, step, name]);

  const child: ReactElement<
    ComponentPropsWithoutRef<any>
  > = typeof children !== 'string' && Array.isArray(children)
    ? children[0] : children;

  return status !== TRANSITION_STATE_UNMOUNTED && (
    !unmountOnExit || mountOnEnter
  ) ? cloneElement(child, {
      className: classNames(
        child.props?.className, getClassName()
      ),
      ...rest,
    }) : null;
};

Transition.displayName = 'Transition';

export default Transition;
