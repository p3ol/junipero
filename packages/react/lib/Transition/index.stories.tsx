import { useReducer } from 'react';
import { mockState } from '@junipero/core';

import Button from '../Button';
import Label from '../Label';
import TextField, { TextFieldChangeEvent } from '../TextField';
import Transition from '.';

export default { title: 'react/Transition' };

export const basic = () => {
  const [state, dispatch] = useReducer(mockState, {
    enabled: false,
    timeout: 100,
  });

  return (
    <>
      <div>
        <Label>Timeout (in ms)</Label>
        <TextField
          value={state.timeout}
          onChange={
            (e: TextFieldChangeEvent) => dispatch({ timeout: Number(e.value) })
          }
        />
      </div>
      <p>
        <Button
          onClick={() => dispatch({ enabled: !state.enabled })}
        >
          Toggle animation
        </Button>
      </p>
      <Transition
        in={state.enabled}
        timeout={state.timeout}
        name="jp-slide-in-down-menu"
      >
        <div>Text</div>
      </Transition>
    </>
  );
};
