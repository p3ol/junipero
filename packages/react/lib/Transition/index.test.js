import { useState } from 'react';
import { fireEvent, render, act } from '@testing-library/react';

import Transition from '.';

describe('<Transition />', () => {
  it('should render correctly', () => {
    jest.useFakeTimers();

    const events = {
      onEnter: jest.fn(),
      onEntering: jest.fn(),
      onEntered: jest.fn(),
      onExit: jest.fn(),
      onExiting: jest.fn(),
      onExited: jest.fn(),
    };

    const Comp = () => {
      const [enabled, setEnabled] = useState(false);

      return (
        <>
          <button onClick={() => setEnabled(e => !e)}>Toggle</button>
          <Transition
            in={enabled}
            timeout={100}
            { ...events }
          >
            <div>Animated text</div>
          </Transition>
        </>
      );
    };

    const { container, unmount } = render(<Comp />);
    expect(container).toMatchSnapshot('Idle');

    // Enter
    fireEvent.click(container.querySelector('button'));
    expect(events.onEnter).toHaveBeenCalledTimes(1);
    act(() => { jest.advanceTimersByTime(50); });
    expect(events.onEntering).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot('Active');
    act(() => { jest.advanceTimersByTime(100); });
    expect(events.onEntered).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot('Done');

    // Exit
    fireEvent.click(container.querySelector('button'));
    expect(events.onExit).toHaveBeenCalledTimes(1);
    act(() => { jest.advanceTimersByTime(50); });
    expect(events.onExiting).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot('Active');
    act(() => { jest.advanceTimersByTime(100); });
    expect(events.onExited).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot('Done');
    unmount();
  });
});
