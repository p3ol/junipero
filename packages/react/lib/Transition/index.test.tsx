import { useState } from 'react';
import { fireEvent, render, act } from '@testing-library/react';
import { vi } from 'vitest';

import Transition from '.';

describe('<Transition />', () => {
  it('should render correctly', () => {
    vi.useFakeTimers();

    const events = {
      onEnter: vi.fn(),
      onEntering: vi.fn(),
      onEntered: vi.fn(),
      onExit: vi.fn(),
      onExiting: vi.fn(),
      onExited: vi.fn(),
    };

    const Comp = () => {
      const [enabled, setEnabled] = useState(false);

      return (
        <>
          <button onClick={() => setEnabled(e => !e)}>Toggle</button>
          <Transition
            in={enabled}
            timeout={100}
            { ...events }
          >
            <div>Animated text</div>
          </Transition>
        </>
      );
    };

    const { container, unmount } = render(<Comp />);
    expect(container).toMatchSnapshot('Idle');

    // Enter
    fireEvent.click(container.querySelector('button'));
    expect(events.onEnter).toHaveBeenCalledTimes(1);
    act(() => { vi.advanceTimersByTime(50); });
    expect(events.onEntering).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot('Active');
    act(() => { vi.advanceTimersByTime(100); });
    expect(events.onEntered).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot('Done');

    // Exit
    fireEvent.click(container.querySelector('button'));
    expect(events.onExit).toHaveBeenCalledTimes(1);
    act(() => { vi.advanceTimersByTime(50); });
    expect(events.onExiting).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot('Active');
    act(() => { vi.advanceTimersByTime(100); });
    expect(events.onExited).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot('Done');
    unmount();
  });
});
