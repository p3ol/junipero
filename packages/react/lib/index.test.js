import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
