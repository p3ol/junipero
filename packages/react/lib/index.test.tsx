import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
