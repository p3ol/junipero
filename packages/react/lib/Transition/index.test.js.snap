// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Transition /> should render correctly 1`] = `
<div>
  <button>
    Toggle
  </button>
</div>
`;

exports[`<Transition /> should render correctly 2`] = `
<div>
  <button>
    Toggle
  </button>
  <div
    class="transition-enter"
  >
    Animated text
  </div>
</div>
`;
